include_core/OMR/TypeTraits.hpp

```diff
@@ -0,0 +1,161 @@
+};
+
+template <typename T, T value>
+const T IntegralConstant<T, value>::VALUE;
+
+template <bool value>
+struct BoolConstant : IntegralConstant<bool, value> {};
+
+struct FalseConstant : BoolConstant<false> {};
+
+struct TrueConstant : BoolConstant<true> {};
+
+template <typename T>
+struct TypeAlias
+{
+	typedef T Type;
+};
+
+///
+/// Type modifications: add or remove qualifiers or type modifiers
+///
+
+/// RemoveConst<T> is T without any const qualifiers.
+template <typename T>
+struct RemoveConst : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveConst<const T> : RemoveConst<T> {};
+
+/// RemoveVolatile<T>::Type is T without volatile qualifiers.
+template <typename T>
+struct RemoveVolatile : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveVolatile<volatile T> : RemoveVolatile<T> {};
+
+/// RemoveCv<T>::Type is T without const/volatile qualifiers.
+template <typename T>
+struct RemoveCv : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveCv<const T> : RemoveCv<T> {};
+
+template <typename T>
+struct RemoveCv<volatile T> : RemoveCv<T> {};
+
+template <typename T>
+struct RemoveCv<const volatile T> : RemoveCv<T> {};
+
+/// RemoveReference<T>::Type is T as a non-reference.
+template <typename T>
+struct RemoveReference : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveReference<T&> : RemoveReference<T> {};
+
+// TODO: Handle RValue references in typetraits
+// template <typename T>
+// struct RemoveReference<T&&> : RemoveReference<T> {};
+
+/// Remove cv-qualifiers and one layer of references.
+template <typename T>
+struct RemoveCvRef : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveCvRef<const T> : RemoveCvRef<T> {};
+
+template <typename T>
+struct RemoveCvRef<volatile T> : RemoveCvRef<T> {};
+
+template <typename T>
+struct RemoveCvRef<const volatile T> : RemoveCvRef<T> {};
+
+template <typename T>
+struct RemoveCvRef<T&> : RemoveCvRef<T> {};
+
+// TODO: Handle rvalue references in type traits
+// template <typename T>
+// struct RemoveCvRef<T&&> : RemoveCvRef<T> {};
+
+///
+/// Type reflection: statically query types
+///
+
+/// True if T and U are the exact same type. Does not remove qualifiers.
+template <typename T, typename U>
+struct IsSame : FalseConstant {};
+
+template <typename T>
+struct IsSame<T, T> : TrueConstant {};
+
+/// true if T is a reference type.
+template <typename T>
+struct IsReference : FalseConstant {};
+
+template <typename T>
+struct IsReference<T&> : TrueConstant {};
+
+// TODO: Handle rvalue references in type traits
+// template <typename T>
+// struct IsReference<T&&> : TrueConstant {};
+
+/// Helper for IsPointer.
+template <typename T>
+struct IsPointerBase : FalseConstant {};
+
+template <typename T>
+struct IsPointerBase<T*> : TrueConstant {};
+
+/// IsPointer<T>::VALUE is true if T is a (possibly cv qualified) primitive pointer type.
+template <typename T>
+struct IsPointer : IsPointerBase<typename RemoveCvRef<T>::Type> {};
```

Comment:
Interesting. This actually deviates from `std::is_pointer`, which only remove cv qualifiers not ref.

Reply:
Hah good eye. This was deliberate, but after some offline talk today, I'll probably change this to match the standard type traits.

Reply:
I stopped treating `T*&` as pointers, and updated the tests to reflect that.